src/addons/hwtoken/src/main/java/ee/ria/xroad/signer/tokenmanager/module/HardwareModuleWorker.java

```diff
@@ -119,21 +120,25 @@ protected void deinitializeModule() throws Exception {
 
         log.info("Module '{}' got {} slots", module.getType(), slots.length);
 
-        Map<String, TokenType> tokens = new HashMap<>();
+        // HSM slot ids defined in module data
+        Set<String> slotIds = module.getSlotIds();
+
+        log.info("Slot configuration for module '{}' defined as {}", module.getType(), slotIds.toString());
 
+        Map<String, TokenType> tokens = new HashMap<>();
         for (int slotIndex = 0; slotIndex < slots.length; slotIndex++) {
             TokenType token = createToken(slots, slotIndex);
-            TokenType previous = tokens.putIfAbsent(token.getId(), token);
-
-            if (previous == null) {
-                log.info("Module '{}' slot #{} has token with ID '{}': {}", module.getType(), slotIndex, token.getId(),
-                        token);
-            } else {
-                log.info("Module '{}' slot #{} has token with ID '{}' but token with that ID is already registered",
-                        module.getType(), slotIndex, token.getId());
+            if (slotIds.isEmpty() || slotIds.contains(token.getId())) {
```

Comment:
The token Id is not the same as the slotId. One should avoid unnecessarily creating the token by checking that the `slot[slotIndex].getSlotId()` is in the slotIds to scan (or slotIds is empty) before creating the token.

(see http://docs.oasis-open.org/pkcs11/pkcs11-base/v2.40/os/pkcs11-base-v2.40-os.html#_Toc72656028)